### PR TITLE
Add timezone_hour Presto function

### PR DESCRIPTION
**Pull request by gopukrishnasIBM**
Resolves https://github.com/facebookincubator/velox/issues/4469

### CODE REVIEW

**Comment by ashokku2022** [comment]
@aditi-pandit Please review when you get time, thanks!

File: velox/docs/functions/presto/datetime.rst [hunk 1]
```diff
@@ -219,3 +219,7 @@ This behavior is different from Presto Java that allows arbitrary large timestam
 .. function:: yow(x) -> bigint
 
     This is an alias for :func:`year_of_week`.
+
+.. function:: timezone_hour(timestamp) -> bigint
```
Comment (mbasmanova) [inline velox/docs/functions/presto/datetime.rst:223]: Please, move this function to preserve alphabetical order.

File: velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp [hunk 2]
```diff
@@ -2837,3 +2837,28 @@ TEST_F(DateTimeFunctionsTest, dateParse) {
   VELOX_ASSERT_THROW(
       dateParse("116", "%y+"), "Invalid format: \"116\" is malformed at \"6\"");
 }
+
+TEST_F(DateTimeFunctionsTest, timeZoneHourTest) {
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:2841]: naming: drop 'Test' suffix from the test method name

File: velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp [hunk 3]
```diff
@@ -2837,3 +2837,28 @@ TEST_F(DateTimeFunctionsTest, dateParse) {
...
+  EXPECT_EQ(5, timezone_hour("2023-01-01 03:20:00", "Asia/Kolkata"));
+  EXPECT_EQ(5, timezone_hour("2023-06-01 03:20:00", "Asia/Kolkata"));
+  // America/Los_Angeles - Day light savings starts from March 12
+  EXPECT_EQ(-8, timezone_hour("2023-03-11 12:00:00", "America/Los_Angeles"));
+  EXPECT_EQ(-7, timezone_hour("2023-03-13 12:00:00", "America/Los_Angeles"));
+  // Different time with same date
+  EXPECT_EQ(-4, timezone_hour("2023-01-01 03:20:00", "Canada/Atlantic"));
+  EXPECT_EQ(-4, timezone_hour("2023-01-01 10:00:00", "Canada/Atlantic"));
+  // Invalid inputs
+  EXPECT_THROW(
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:2861]: Use VELOX_ASSERT_THROW

**Review by mbasmanova — COMMENTED** [review]
@Yuhta Jimmy, would you help review this PR?

File: velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp [hunk 4]
```diff
@@ -2837,3 +2837,31 @@ TEST_F(DateTimeFunctionsTest, dateParse) {
...
+                    "timezone_hour(c0)", timestamp, timezone)
+                    .value();
+    return hour;
+  };
+
+  // Asia/Kolkata - should return 5 throughout the year
+  EXPECT_EQ(5, timezone_hour("2023-01-01 03:20:00", "Asia/Kolkata"));
+  EXPECT_EQ(5, timezone_hour("2023-06-01 03:20:00", "Asia/Kolkata"));
+  // America/Los_Angeles - Day light savings starts from March 12
+  EXPECT_EQ(-8, timezone_hour("2023-03-11 12:00:00", "America/Los_Angeles"));
```
Comment (Yuhta) [inline velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:2855]: What do we get for `timezone_hour("2023-03-12 02:30:00", "America/Los_Angeles")` and `timezone_hour("2023-11-15 01:30:00", "America/Los_Angeles")`?

Comment (gopukrishnasIBM) [reply velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:2855]: @Yuhta Am getting the following values.
timezone_hour("2023-03-12 02:30:00", "America/Los_Angeles") -> -8
timezone_hour("2023-11-15 01:30:00", "America/Los_Angeles") -> -8
Do I need to add these 2 also in test case?

Comment (Yuhta) [reply velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:2855]: Can you check these 2 are compatible with Presto Java behavior?  Once confirmed we can add these 2 to unit test.

Comment (gopukrishnasIBM) [reply velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:2855]: @Yuhta Confirmed with Java behavior.
image

Added these 2 to unit test.

Comment (Yuhta) [reply velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:2855]: Oops, I think I made a typo, the second one should be 2023-11-05 01:30:00, can you check this one?

Once confirmed this should be good.  There are also some merge conflicts to resolve, then we can merge this.

Comment (gopukrishnasIBM) [reply velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:3025]: @Yuhta For 2023-11-05 01:30:00 am getting -7, which is the same in Java too.
image

Added that case also in unit test.

**Comment by gopukrishnasIBM** [comment]
@Yuhta @mbasmanova Is there anything else I could do here to close the PR?

File: velox/functions/prestosql/DateTimeFunctions.h [hunk 5]
```diff
@@ -1041,4 +1041,30 @@ struct ParseDateTimeFunction {
   }
 };
 
+template <typename T>
+struct TimeZoneHourFunction : public TimestampWithTimezoneSupport<T> {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE void call(
+      int64_t& result,
+      const arg_type<TimestampWithTimezone>& timestampWithTimezone) {
```
[the remark this answers is not in the document]
Comment (gopukrishnasIBM) [reply velox/functions/prestosql/DateTimeFunctions.h]: @aditi-pandit Parameter name changed to input.

**Pull request closed without merging**